src/js/node/events.js

```diff
@@ -92,9 +92,11 @@ const emitWithoutRejectionCapture = function emit(type, ...args) {
   if (events === undefined) return false;
   var handlers = events[type];
   if (handlers === undefined) return false;
-
-  for (var handler of [...handlers]) {
-    handler.apply(this, args);
+  // iterate in reverse so the handlers array does not need to be cloned
+  for (let i = handlers.length - 1; i >= 0; --i) {
+    // for performance reasons Function.call(...) is used if only one arg needs to be passed
+    if (args.length === 1) handlers[i].call(this, args[0]);
+    else handlers[i].apply(this, args);
   }
   return true;
```

Comment:
The big perf win is the `.call` when args are 0 or 1. It can be done to 0, 1, 2 (or however a common call is) before going to `apply`

Reply:
Yes `.apply()` is super slow for some reason. I now use `.call()` in the cases for 0, 1, 2 and 3 args and `.apply()` else.

Here is a quick bench for the case with `args.length = 1`:
```
handler.apply()                      time (avg)             (min … max)       p75       p99      p995
----------------------------------------------------------------------- -----------------------------
single emit (one handler): bun    69.65 ns/iter  (57.65 ns … 337.51 ns)  61.58 ns  299.2 ns 306.25 ns
single emit (two handlers): bun  124.88 ns/iter (107.02 ns … 377.68 ns)  117.1 ns 343.13 ns 355.99 ns
on x 10_000 (one handler): bun    10.01 ms/iter     (3.9 ms … 17.88 ms)  12.75 ms  17.88 ms  17.88 ms
on x 10_000 (two handlers): bun   15.82 ms/iter    (7.33 ms … 24.57 ms)  19.88 ms  24.57 ms  24.57 ms


handler.call()                       time (avg)             (min … max)       p75       p99      p995
----------------------------------------------------------------------- -----------------------------
single emit (one handler): bun     42.9 ns/iter  (32.83 ns … 450.39 ns)  34.58 ns 250.83 ns 267.58 ns
single emit (two handlers): bun   69.18 ns/iter  (53.71 ns … 351.81 ns)  63.47 ns  295.4 ns 322.01 ns
on x 10_000 (one handler): bun      5.2 ms/iter    (1.63 ms … 10.01 ms)   6.87 ms   9.61 ms  10.01 ms
on x 10_000 (two handlers): bun     7.7 ms/iter    (2.68 ms … 13.66 ms)   10.3 ms  13.66 ms  13.66 ms
```

Node.js seems to always use `apply()` though with no performance loss, as far as I can tell. Could be an issue in JavaScriptCore? Or where is apply implemented?

Reply:
It's been an issue, `call` vs `apply`, in the past with engines so it's definitely possible.

Reply:
i just want to add, and it may not perform well but it is worth trying: `call(x, ...args)` instead of apply. i'm sure they both are jit-d the same but just throwing out there.

Reply:
I just tried that and spreading the args into `.call` was way slower than using `.apply`. Not sure what is causing that exactly.

#### For a test with 4 args in the emit
```
handler.call(this, ...args)        time (avg)             (min … max)       p75       p99      p995
--------------------------------------------------------------------- -----------------------------
single emit (5 handlers): bun  274.39 ns/iter (204.39 ns … 634.24 ns) 321.03 ns 587.95 ns 634.24 ns
single emit (6 handlers): bun  313.08 ns/iter  (240.1 ns … 568.57 ns) 414.08 ns 551.74 ns 568.57 ns
single emit (7 handlers): bun  362.51 ns/iter (274.87 ns … 593.25 ns)  453.3 ns 572.99 ns 593.25 ns
single emit (8 handlers): bun  387.85 ns/iter (305.97 ns … 637.27 ns) 496.41 ns  587.6 ns 637.27 ns
on x 10_000 (5 handlers): bun   54.83 ms/iter   (40.74 ms … 70.82 ms)  65.17 ms  70.82 ms  70.82 ms
on x 10_000 (6 handlers): bun   64.24 ms/iter    (40.2 ms … 90.42 ms)  78.42 ms  90.42 ms  90.42 ms
on x 10_000 (7 handlers): bun   73.53 ms/iter   (45.6 ms … 101.21 ms)  86.55 ms 101.21 ms 101.21 ms
on x 10_000 (8 handlers): bun   84.14 ms/iter  (52.45 ms … 109.17 ms) 100.06 ms 109.17 ms 109.17 ms

handler.apply(this, args)          time (avg)             (min … max)       p75       p99      p995
--------------------------------------------------------------------- -----------------------------
single emit (5 handlers): bun  144.27 ns/iter (119.34 ns … 617.92 ns) 131.89 ns 376.44 ns 401.73 ns
single emit (6 handlers): bun  183.05 ns/iter (163.72 ns … 441.73 ns) 173.29 ns 405.11 ns 423.69 ns
single emit (7 handlers): bun  198.07 ns/iter (173.14 ns … 451.08 ns)  193.1 ns 408.01 ns 422.04 ns
single emit (8 handlers): bun  216.75 ns/iter (193.62 ns … 470.83 ns) 203.98 ns 448.03 ns 450.76 ns
on x 10_000 (5 handlers): bun   37.06 ms/iter   (20.85 ms … 52.62 ms)  45.94 ms  52.62 ms  52.62 ms
on x 10_000 (6 handlers): bun   43.45 ms/iter    (24.4 ms … 61.58 ms)  52.84 ms  61.58 ms  61.58 ms
on x 10_000 (7 handlers): bun   47.95 ms/iter   (29.53 ms … 67.53 ms)  58.41 ms  67.53 ms  67.53 ms
on x 10_000 (8 handlers): bun   54.75 ms/iter   (33.73 ms … 74.82 ms)  65.29 ms  74.82 ms  74.82 ms
```